iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,369 @@
+                    Ok(message) => MessageResult::new_message(message),
+                    Err(e) => MessageResult::new_error(e)
+                };
+
+                yield result;
+            }
+        }
+    }
+
+    async fn handshake(&mut self) -> Result<(), io::Error> {
+        match &self.state {
+            State::Connecting => {
+                // Connecting to peer
+                let addr = self.id.address.clone();
+                let stream = TcpStream::connect(addr.clone()).await;
+                match stream {
+                    Ok(stream) => {
+                        let (read, write) = stream.into_split();
+                        self.read = Some(read);
+                        self.write = Some(write);
+                        self.state = State::ConnectedTo;
+                    }
+                    Err(e) => {
+                        warn!("Could not connect to peer on {}! {}", addr, e);
+                    }
+                }
+            }
+            State::ConnectedTo => {
+                // Send client hello with out public key
+                #[allow(clippy::expect_used)]
+                let mut write_half = self.write.as_mut().expect("Unexpected");
+                write_half.as_ref().writable().await?;
+                send_client_hello(&mut write_half, self.public_key.0.as_slice()).await?;
+                // Read server hello with node's public key
+                #[allow(clippy::expect_used)]
+                let read_half = self.read.as_mut().expect("Unexpected");
+                let public_key = read_server_hello(read_half).await?;
+                let x25519 = X25519Sha256::new();
+                let shared = match x25519.compute_shared_secret(&self.secret_key, &public_key) {
+                    Ok(key) => key,
+                    Err(e) => {
+                        warn!("Error creating shared secret! {}", e);
+                        return Err(io::Error::from(ErrorKind::InvalidInput));
+                    }
+                };
+                self.cipher = Some(self.new_encryptor(shared.0.as_slice()));
+                self.state = State::Ready;
+            }
+            State::ConnectedFrom => {
+                // Reading client public key from client hello,
+                // creating shared secret and sending our public key to client
+                #[allow(clippy::expect_used)]
+                let read_half = self.read.as_mut().expect("Unexpected");
+                let public_key = read_client_hello(read_half).await?;
+                let x25519 = X25519Sha256::new();
+                let shared = match x25519.compute_shared_secret(&self.secret_key, &public_key) {
+                    Ok(key) => key,
+                    Err(e) => {
+                        warn!("Error creating shared secret! {}", e);
+                        return Err(io::Error::from(ErrorKind::InvalidInput));
+                    }
+                };
```

Comment:
There is a repetition here. May be you can abstract it somehow

Reply:
Done, extracted a method.